numba/core/types/functions.py

```diff
@@ -442,7 +441,10 @@ def get_call_type(self, context, args, kws):
                                 raise exc
                         nonliteral_e = exc
                     else:
-                        break
+                        # If out returns None don't break as we may need to use
+                        # a literal.
+                        if out is not None:
+                            break
```

Comment:
I think the typing for the `@bound_function("array.astype")` should raise a `numba.core.errors.RequireLiteralValue` exception if a `unicode_type` appears, this would be consistent with how e.g. `@overload` would handle this situation and doesn't require another code path here.

Reply:
This change worked. I've reverted the changes to `numba/core/types/functions.py` and updated the bound function. Thank you.

Reply:
Great, thanks.